edx_oauth_client/middleware.py

```diff
@@ -1,61 +1,96 @@
+                expires="Thu, 01-Jan-1970 00:00:00 GMT",
+            )
+
+        return response
 
     def process_request(self, request):
         """
-        Check multidomain cookie and if user is authenticated on sso, login it on edx.
+        Checks cross-domain cookies and, if the user is authenticated with SSO,
+        authorizes the user on edX.
         """
-        backend = "edx-oauth2"
+        backend = GenericOAuthBackend.name
         current_url = request.get_full_path()
 
-        # don't work for admin
-        for attr in ['SOCIAL_AUTH_EXCLUDE_URL_PATTERN', 'AUTOCOMPLETE_EXCLUDE_URL_PATTERN']:
-            if hasattr(settings, attr):
-                r = re.compile(getattr(settings, attr))
-                if r.match(current_url):
-                    return None
-
-        auth_cookie = request.COOKIES.get(self.cookie_name, '0').lower()
-        auth_cookie_user = request.COOKIES.get('{}_user'.format(self.cookie_name))
-        auth_cookie = (auth_cookie in ('1', 'true', 'ok'))
-        continue_url = reverse('{0}:complete'.format(NAMESPACE),
-                               args=(backend,))
+        # SeamlessAuthorization doesn't work for Django administration
+        if hasattr(settings, "SOCIAL_AUTH_EXCLUDE_URL_PATTERN"):
+            r = re.compile(settings.SOCIAL_AUTH_EXCLUDE_URL_PATTERN)
+            if r.match(current_url):
+                return None
+
+        auth_cookie = request.COOKIES.get(self.cookie_name)
+        auth_cookie_portal = request.session.get(self.cookie_name)
+        continue_url = reverse("{0}:complete".format(NAMESPACE), args=(backend,))
         is_auth = request.user.is_authenticated()
+        is_same_user = (auth_cookie == auth_cookie_portal)
 
-        is_same_user = (request.user.username == auth_cookie_user)
+        # Check for infinity redirection loop
+        is_continue = (continue_url in current_url)
 
-        OAUTH_PROVIDER_URL = settings.FEATURES.get("OAUTH_PROVIDER_URL", "")
+        request.session[self.cookie_name] = auth_cookie
 
         if reverse("logout") == current_url:
+            del request.session[self.cookie_name]
             logout(request)
-            return redirect(urljoin(OAUTH_PROVIDER_URL, '/logout'))
+            # FIXME: redirect is dirty fix to be able logout external portal.
+            # Please try to avoid and don't do it at edx side.
+            return redirect(urljoin(self.provider_url, self.provider_logout_url_path))
 
-        # Check for infinity redirection loop
-        is_continue = (continue_url in current_url)
+        if not is_same_user and is_auth:
+            logout(request)
 
-        if (auth_cookie and not is_continue and (not is_auth or not is_same_user)) or \
-                ('force_auth' in request.session and request.session.pop('force_auth')):
+        if (auth_cookie and not is_continue and (not is_auth or not is_same_user)) or (
+            "force_auth" in request.session and request.session.pop("force_auth")
```

Comment:
The condition could be decreased:
```suggestion
           request.session.pop("force_auth", None)
```